File: examples/java/src/main/java/runar/examples/stateful-wots-gate/StatefulWOTSGate.runar.java
package runar.examples.statefulwotsgate;

import runar.lang.StatefulSmartContract;
import runar.lang.annotations.Public;
import runar.lang.types.Bigint;
import runar.lang.types.ByteString;
import static runar.lang.Builtins.assertThat;
import static runar.lang.Builtins.verifyWOTS;

// StatefulWOTSGate — stateful + post-quantum interaction fixture (GAP-407).
class StatefulWOTSGate extends StatefulSmartContract {
    Bigint count;

    StatefulWOTSGate(Bigint count) {
        super(count);
        this.count = count;
    }

    @Public
    void advance(ByteString msg, ByteString sig, ByteString wotsPubKey) {
        assertThat(verifyWOTS(msg, sig, wotsPubKey));
        this.count = this.count.plus(Bigint.ONE);
    }
}

File: examples/ts/stateful-wots-gate/StatefulWOTSGate.runar.ts
import { StatefulSmartContract, assert, verifyWOTS } from 'runar-lang';
import type { ByteString } from 'runar-lang';

// StatefulWOTSGate - stateful + post-quantum interaction fixture (GAP-407).
// A counter whose state mutation is gated on a WOTS+ one-time signature,
// composing the auto-injected stateful covenant (checkPreimage + 0x41 sighash
// pin + state continuation) with a ~10 KB post-quantum verify in one script.
class StatefulWOTSGate extends StatefulSmartContract {
  count: bigint;

  constructor(count: bigint) {
    super(count);
    this.count = count;
  }

  public advance(msg: ByteString, sig: ByteString, wotsPubKey: ByteString) {
    assert(verifyWOTS(msg, sig, wotsPubKey));
    this.count = this.count + 1n;
  }
}
